From: Hannes Reinecke <hare@suse.de>
Date: Mon, 15 Feb 2016 11:31:51 +0100
Subject: [PATCH] scsi_dh_alua: Do not block request queue if workqueue is active
References: bsc#960458
Patch-Mainline: n/a; alua update has been reworked for mainline

If an item is queued to the alua workqueue we should not block the
queue, as this item might be queued due to an I/O returning with
a unit attention code. In these cases the I/O from the alua handler
will be queued _after_ the original I/O, causing an I/O stall as
the original I/O can never be completed.

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 drivers/scsi/device_handler/scsi_dh_alua.c | 6 +-----
 1 file changed, 1 insertion(+), 5 deletions(-)

diff --git a/drivers/scsi/device_handler/scsi_dh_alua.c b/drivers/scsi/device_handler/scsi_dh_alua.c
index f6480a6..471a8b2 100644
--- a/drivers/scsi/device_handler/scsi_dh_alua.c
+++ b/drivers/scsi/device_handler/scsi_dh_alua.c
@@ -1201,12 +1201,8 @@ static int alua_prep_fn(struct scsi_device *sdev, struct request *req)
 
 	rcu_read_lock();
 	pg = rcu_dereference(h->pg);
-	if (pg) {
+	if (pg)
 		state = pg->state;
-		/* Defer I/O while rtpg_work is active */
-		if (pg->rtpg_sdev)
-			state = TPGS_STATE_TRANSITIONING;
-	}
 	rcu_read_unlock();
 	if (state == TPGS_STATE_TRANSITIONING)
 		ret = BLKPREP_DEFER;
-- 
1.8.5.6

